conops/simulation/acs.py

```
@@ -635,6 +1037,1175 @@ def _calculate_safe_mode_pointing(self, utime: float) -> None:
+                if r_sun_mag > 0:
+                    AU = 1.495978707e11
+                    P0 = 4.56e-6  # N/m^2 at 1 AU
+                    scale = (AU / r_sun_mag) ** 2
+                    r_sun_body = self._inertial_to_body(r_sun)
+                    sun_hat = r_sun_body / np.linalg.norm(r_sun_body)
+                    F_srp = (
+                        P0
+                        * scale
+                        * self._solar_area
+                        * self._solar_reflectivity
+                        * sun_hat
+                    )
+                    torque_srp = np.cross(self._cp_offset, F_srp)
+                    torque += torque_srp
+                    srp_mag = float(np.linalg.norm(torque_srp))
+            except Exception:
+                pass
+
+        # Magnetic torque from residual moment
+        if np.linalg.norm(self._residual_moment) > 0:
+            B_body, _ = self._local_bfield_vector(utime)
+            torque_mag = np.cross(self._residual_moment, B_body)
+            torque += torque_mag
+            mag_mag = float(np.linalg.norm(torque_mag))
+
+        # Store last components for telemetry
+        self._last_disturbance_components = {
+            "total": float(np.linalg.norm(torque)),
+            "gg": gg_mag,
+            "drag": drag_mag,
+            "srp": srp_mag,
+            "mag": mag_mag,
+            "vector": torque.tolist(),
+        }
+
+        return torque
+
+    def _atmospheric_density(
+        self, utime: float, alt_m: float, lat_deg: float | None, lon_deg: float | None
+    ) -> float:
+        """Lookup/interpolate atmospheric density (kg/m^3) vs altitude.
+
+        If use_msis_density is enabled and pymsis is available, falls back to MSIS.
+        """
+        # Try MSIS if requested
+        if self._use_msis:
+            try:
+                from datetime import datetime as _dt
+
+                from pymsis import msise00  # type: ignore[import-untyped]
```

Not sure what the intent here of failing silently if pymsis is not installed. Remove this try/except and just add it to pyproject.toml as a dependency.

Move these imports to the top of the file. Doing `_dt` is weird, just do `from datetime import datetime`.